### PR TITLE
Add wildcardSearch ability to elasticsearch (#1148)

### LINKED ISSUE #1148: Allow wildcard service search for elasticsearch storage
## Requirement - what kind of business use case are you trying to solve?

We have a need to collate traces across different services that share a specific tag value. The traces are initiated by separate HTTP requests and therefore appear as separate traces under separate services. We would like to search by a tag across all services with all matches displaying regardless of service name. 

## Problem - what in Jaeger blocks you from solving the requirement?

The jaeger-ui and the jaeger-query modules require service as a parameter. 

## Proposal - what do you suggest to solve the problem or improve the existing situation?

The jaeger-ui should include an 'All' or 'Any' option in the services dropdown. 

image

The jaeger-query module should handle a request containing the value `service=all` or `service=any` by creating a new wildcard query. The `elastic.NewWildcardQuery` is supported by the `bool.Must` query builder. 

## Any open questions to address

I have a working example with tests, but would like to discuss prior to opening a pull request.

### PR DESCRIPTION
**Pull request by ryansos**
Resolves #1148 

Signed-off-by: Ryan Dwyer <rdwyer@rapidsos.com>

## Which problem is this PR solving?
- allows users to search by wildcard against the serviceName in elasticsearch

## Short description of the changes
- adds a config variable WildcardSearch (bool) which defaults to false
- this will require changes to the jaeger-ui repo (pr to follow)

### CODE REVIEW

**Comment by pavolloffay** [comment]
Yes it is look at `es-integration-test.sh`

```
docker run --rm -it -p 9200:9200 -e "http.host=0.0.0.0" -e "transport.host=127.0.0.1" docker.elastic.co/elasticsearch/elasticsearch:5.4.0
STORAGE=elasticsearch make storage-integration-test
```

**Comment by pavolloffay** [comment]
Wondering whether the ES failure is related to this PR or #1108

**Comment by ryansos** [comment]
> Wondering whether the ES failure is related to this PR or #1108

interesting - it appears to have passed.

**Comment by pavolloffay** [comment]
yah, I think the failure is related to #1108

File: cmd/query/app/query_parser.go [hunk 1]
```diff
@@ -29,23 +29,26 @@ import (
...
-	tagsParam        = "tags"
-	startTimeParam   = "start"
-	limitParam       = "limit"
-	minDurationParam = "minDuration"
-	maxDurationParam = "maxDuration"
-	serviceParam     = "service"
-	endTimeParam     = "end"
-	prettyPrintParam = "prettyPrint"
+	defaultQueryLimit     = 100
+	defaultAnyServiceName = "*"
```
Comment (yurishkuro) [inline cmd/query/app/query_parser.go]: I think this makes more sense as a boolean flag, similar to prettyPrint.
```go
	if prettyPrint := r.FormValue(prettyPrintParam); prettyPrint != "" && prettyPrint != "false" {}
```

File: cmd/query/app/query_parser.go [hunk 2]
```diff
@@ -79,6 +82,7 @@ type traceQueryParameters struct {
 //     tags :== 'tags=' jsonMap
 func (p *queryParser) parse(r *http.Request) (*traceQueryParameters, error) {
 	service := r.FormValue(serviceParam)
+	anyServiceName := r.FormValue(anyServiceNameParam)
```
Comment (yurishkuro) [inline cmd/query/app/query_parser.go]: please add to the comments of this method

File: storage/spanstore/interface.go [hunk 3]
```diff
@@ -42,12 +42,13 @@ type Reader interface {
...
-	ServiceName   string
-	OperationName string
-	Tags          map[string]string
-	StartTimeMin  time.Time
-	StartTimeMax  time.Time
-	DurationMin   time.Duration
-	DurationMax   time.Duration
-	NumTraces     int
+	ServiceName    string
+	AnyServiceName string
```
Comment (yurishkuro) [inline storage/spanstore/interface.go]: please make it Boolean

File: pkg/es/config/config.go [hunk 4]
```diff
@@ -57,6 +58,7 @@ type ClientBuilder interface {
 	GetTagsFilePath() string
 	GetAllTagsAsFields() bool
 	GetTagDotReplacement() string
+	GetWildcardSearch() bool
```
Comment (yurishkuro) [inline pkg/es/config/config.go]: s/GetWildcardSearch/IsWildcardSearch/  (GetAllTagsAsFields was badly named)

File: plugin/storage/es/options.go [hunk 5]
```diff
@@ -41,6 +41,7 @@ const (
 	suffixTagsAsFieldsAll   = suffixTagsAsFields + ".all"
 	suffixTagsFile          = suffixTagsAsFields + ".config-file"
 	suffixTagDeDotChar      = suffixTagsAsFields + ".dot-replacement"
+	suffixWildcardSearch    = ".wildcard-search"
```
Comment (yurishkuro) [inline plugin/storage/es/options.go]: this name seems overly broad. Are you really supporting wildcard search for everything, or just for the service name?

File: plugin/storage/es/spanstore/reader.go [hunk 6]
```diff
@@ -421,12 +425,17 @@ func (s *SpanReader) buildFindTraceIDsQuery(traceQuery *spanstore.TraceQueryPara
 	startTimeQuery := s.buildStartTimeQuery(traceQuery.StartTimeMin, traceQuery.StartTimeMax)
 	boolQuery.Must(startTimeQuery)
 
-	//add process.serviceName query
-	if traceQuery.ServiceName != "" {
+	//add process.serviceName query, only if AnyServiceName is blank
+	if traceQuery.ServiceName != "" && !traceQuery.AnyServiceName {
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/reader.go]: I would prefer to return an error when both `traceQuery.ServiceName` and `traceQuery.AnyServiceName` are non-empty

File: plugin/storage/es/spanstore/reader.go [hunk 7]
```diff
@@ -459,6 +468,10 @@ func (s *SpanReader) buildServiceNameQuery(serviceName string) elastic.Query {
 	return elastic.NewMatchQuery(serviceNameField, serviceName)
 }
 
+func (s *SpanReader) buildAnyServiceNameQuery() elastic.Query {
+	return elastic.NewWildcardQuery(serviceNameField, defaultAnyServiceName)
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/reader.go]: I suggest removing `defaultAnyServiceName` constant and using `*` explicitly here.

File: plugin/storage/factory_config.go [hunk 8]
```diff
@@ -36,6 +36,7 @@ type FactoryConfig struct {
 	SpanWriterTypes         []string
 	SpanReaderType          string
 	DependenciesStorageType string
+	WildcardSearch          bool
```
Comment (yurishkuro) [inline plugin/storage/factory_config.go]: is this needed at the generic factory level?

**Comment by jpkrohling** [comment]
@ryansos, are you still interested in this one?

**Pull request closed without merging**